Feature Controls - Reserved Role Apps

kobelb opened the pull request:
# Summary
This has to be run against https://github.com/kobelb/elasticsearch/tree/fc/reserved-role-apps-3 for the time being. I'm waiting on the wildcard support from ES before I create a PR to make these changes, but properly.

This enables [features controls](https://github.com/elastic/kibana/issues/20277) for both the monitoring and machine learning applications. Both of these applications rely on reserved roles to be able to grant access from the security perspective, so the approach differs slightly from the other applications. Using Spaces, it's possible to disable Monitoring and Machine Learning; however, granting access requires the user continue assigning the `monitoring_user`, `machine_learning_user` and `machine_learning_admin` roles. We're disclosing this requirement on the Role management screen with the following:

<img width="1415" alt="Screen Shot 2019-03-22 at 12 27 32 PM" src="https://user-images.githubusercontent.com/627123/54848186-e5c54d00-4c9d-11e9-9b97-54599d5999f3.png">

Emulating the existing behavior prior to feature controls, you'll be required to assign a role which grants access to Kibana, for example `kibana_user` or another custom role, in addition to `monitoring_user` or `machine_learning_user`/`machine_learning_admin` for ML to be visible.

# Monitoring - @chrisronline  or @ycombinator 

If you could please review the code changed under `x-pack/plugins/monitoring`, it would be greatly appreciated. 

We're registering the following feature with x-pack, which enables the Spaces driven feature controls and the security driven feature privileges by registering the following feature during monitoring's [init](https://github.com/kobelb/kibana/blob/7567c9a4c22f888af618b2a8120dc614c2901f9a/x-pack/plugins/monitoring/init.js#L58-L80) lifecycle:

```
xpackMainPlugin.registerFeature({
  id: 'monitoring',
  name: i18n.translate('xpack.monitoring.featureRegistry.monitoringFeatureName', {
    defaultMessage: 'Stack Monitoring',
  }),
  icon: 'monitoringApp',
  navLinkId: 'monitoring',
  app: ['monitoring', 'kibana'],
  catalogue: ['monitoring'],
  privileges: {},
  reserved: {
    privilege: {
      savedObject: {
        all: [],
        read: ['config']
      },
      ui: [],
    },
    description: i18n.translate('xpack.monitoring.feature.reserved.description', {
      defaultMessage: 'To grant users access, you should also assign the monitoring_user role.'
    })
  }
});
```

We're added some functional tests for this change in behavior [here](https://github.com/kobelb/kibana/tree/fc/reserved-role-apps-3/x-pack/test/functional/apps/monitoring/feature_controls)


# Machine Learning - @peteharverson 

If you could please review the code changed under `x-pack/plugins/ml`, it would be greatly appreciated. 

We're registering the following feature with x-pack, which enables the Spaces driven feature controls and the security driven feature privileges by registering the following feature during ML's [init](https://github.com/kobelb/kibana/blob/7567c9a4c22f888af618b2a8120dc614c2901f9a/x-pack/plugins/ml/index.js#L82-L104) lifecycle:

```
xpackMainPlugin.registerFeature({
  id: 'ml',
  name: i18n.translate('xpack.ml.featureRegistry.mlFeatureName', {
    defaultMessage: 'Machine Learning',
  }),
  icon: 'machineLearningApp',
  navLinkId: 'ml',
  app: ['ml', 'kibana'],
  catalogue: ['ml'],
  privileges: {},
  reserved: {
    privilege: {
      savedObject: {
        all: [],
        read: ['config']
      },
      ui: [],
    },
    description: i18n.translate('xpack.ml.feature.reserved.description', {
      defaultMessage: 'To grant users access, you should also assign either the machine_learning_user or machine_learning_admin role.'
    })
  }
});
```

We're added some functional tests for this change in behavior [here](https://github.com/kobelb/kibana/tree/fc/reserved-role-apps-3/x-pack/test/functional/apps/machine_learning)

kobelb committed:
Removing feature privileges from ml/monitoring/apm

kobelb committed:
Adding monitoring/ml/apm as hard-coded global privileges

kobelb committed:
A poorly named abstraction enters the room

kobelb committed:
No more wildcards, starting to move some stuff around

kobelb committed:
Splitting out the feature privilege builders

kobelb committed:
Using actions instead of relying on their implementation

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/no-wildcards

kobelb committed:
We don't need the saved object types any longer

kobelb committed:
Explicitly specifying some actions that used to rely on wildcards

kobelb committed:
Fixing api integration test for privileges

kobelb committed:
Test fixture plugin which adds the globaltype now specifies a feature

kobelb committed:
Unauthorized to find unknown types now

kobelb committed:
Merge branch 'fc/no-wildcards' into fc/reserved/role-apps-3

kobelb committed:
Adding reserved privileges tests

kobelb committed:
Adding reserved privileges in a designated reserved bucket

kobelb committed:
Fixing ui capability tests

kobelb committed:
Adding spaces api tests for apm/ml/monitoring users

kobelb committed:
Adding more roles to the security only ui capability tests

kobelb committed:
You can put a role with reserved privileges using the API

kobelb committed:
Adding support to get roles with _reserved privileges

kobelb committed:
Adding APM functional tests

kobelb committed:
Adding monitoring functional tests

kobelb committed:
Merge branch 'granular-app-privileges' into fc/reserved-role-apps-3

kobelb committed:
Fixing typo

kobelb committed:
Ensuring apm_user, monitoring_user alone don't authorize you

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11972/)

kobelb committed:
Adding ml functional tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12022/)

kobelb committed:
Fixing test

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12030/)

kobelb committed:
Fixing some type errors

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12071/)

kobelb committed:
Updating snapshots

kobelb committed:
Fixing privileges tests

kobelb committed:
Trying to force this to run from source

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12122/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

kobelb committed:
Fixing TS errors

kobelb committed:
Being a less noisy neighbor

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12282/)

kobelb committed:
Forcing logout for apm/dashboard feature controls security tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12343/)

kobelb commented:
retest

kobelb commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12432/)

kobelb committed:
Fixing the security only ui capability tests

kobelb committed:
Removing test that monitoring now tests itself

kobelb commented:
retest

kobelb commented:
retest

spalger commented:
retest

spalger commented:
retest

spalger commented:
retest

spalger commented:
retest

spalger commented:
retest

spalger commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/209/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

kobelb committed:
Fixing some ui capability tests

kobelb committed:
Cleaning up the error page services

kobelb committed:
Fixing misspelling in comment

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/926/)

legrego commented on x-pack/plugins/security/server/lib/authorization/privilege_serializer.ts:
```diff
@@ -29,8 +30,12 @@ export class PrivilegeSerializer {
+
   public static serializeGlobalBasePrivilege(privilegeName: string) {
-    if (!basePrivilegeNames.includes(privilegeName)) {
+    if (!globalBasePrivileges.includes(privilegeName)) {
```
Can you explain why we treat `apm`, `ml`, and `monitoring` like global base privileges?

kobelb replied:
This is left-over code that I hacked in the for the product review, good catch, I'll be removing this.

legrego commented on x-pack/plugins/security/server/routes/api/public/roles/get.js:
```diff
@@ -42,10 +42,17 @@ export function initGetRolesApi(server, callWithRequest, routePreCheckLicenseFn,
-          const featurePrivileges = privileges.filter(privilege => !PrivilegeSerializer.isSerializedGlobalBasePrivilege(privilege));
+          const featurePrivileges = privileges.filter(privilege =>
+            !PrivilegeSerializer.isSerializedGlobalBasePrivilege(privilege) &&
+            !PrivilegeSerializer.isSerializedReservedPrivilege(privilege)
```
If `isSerializedGlobalBasePrivilege` is correct (see my earlier comment), then we shouldn't also need to check `isSerializedReservedPrivilege`, because reserved privileges are being treated like global base privileges (which doesn't feel right IMO)

kobelb committed:
Using forceLogout for monitoring

legrego commented on x-pack/plugins/security/server/routes/api/public/roles/get.js:
```diff
@@ -42,10 +42,17 @@ export function initGetRolesApi(server, callWithRequest, routePreCheckLicenseFn,
       value: roleKibanaApplications.map(({ resources, privileges }) => {
       // if we're dealing with a global entry, which we've ensured above is only possible if it's the only item in the array
         if (resources.length === 1 && resources[0] === GLOBAL_RESOURCE) {
+          const reservedPrivileges = privileges.filter(privilege => PrivilegeSerializer.isSerializedReservedPrivilege(privilege));
```
Should we return a `_transform_error` if we detect reserved privileges assigned at a space resource? If I understand the logic correctly, they will be deserialized as space feature privileges, which I wouldn't expect to happen

kobelb replied:
Yup, I will make it so.

legrego commented on x-pack/test/functional/apps/machine_learning/feature_controls/ml_spaces.ts:
```diff
@@ -0,0 +1,93 @@
+          basePath: '/s/custom_space',
+        });
+
+        await testSubjects.existOrFail('ml-jobs-list', 10000);
```
nit: is the custom timeout necessary here?

kobelb replied:
Nope, just copied code... will remove.

legrego commented on x-pack/test/functional/config.js:
```diff
@@ -81,6 +81,7 @@ export default async function ({ readConfigFile }) {
   return {
     // list paths to the files that contain your plugins tests
     testFiles: [
+      resolve(__dirname, './apps/apm'),
```
nit: apm > advanced settings if you were going for an alpha-sorted list

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/900/)

legrego reviewed (commented):
This is looking really, good, just a couple nits/questions.

I noticed that the reserved roles w/ reserved privileges show incorrectly on the role management screen. It looks like they have a global privilege entry with no privileges assigned. I'm fine merging this PR as-is, but we'll need to at least open a subsequent PR to address the UI. I don't mind taking care of that once I'm back:

![image](https://user-images.githubusercontent.com/3493255/52880887-2f67c880-3131-11e9-9260-9bf3ecacdb1b.png)

kobelb commented on test/scripts/jenkins_xpack_ci_group.sh:
```diff
@@ -39,6 +39,6 @@ tar -xzf "$linuxBuild" -C "$installDir" --strip=1
 echo " -> Running functional and api tests"
 cd "$XPACK_DIR"
-node scripts/functional_tests --debug --bail --kibana-install-dir "$installDir" --include-tag "ciGroup$CI_GROUP"
+node scripts/functional_tests --debug --bail --kibana-install-dir "$installDir" --include-tag "ciGroup$CI_GROUP" --esFrom "source"
```
I can probably remove this, I had CI not run from source during a run or two and forcing this in there made it work.

kobelb replied:
This is unfortunately required at the moment to get CI to run from source, I've talked to Tyler and we're trying to track down where this bug is.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/960/)

kobelb committed:
Removing code that never should have been there, sorry Larry

kobelb committed:
Less leniency with the get roles

kobelb committed:
Barely alphabetical for a bit

legrego commented on x-pack/plugins/security/server/lib/authorization/privilege_serializer.test.ts:
```diff
@@ -36,6 +36,43 @@ describe(`#isSerializedSpaceBasePrivilege`, () => {
+    'feature_foo',
+    'feature_foo.privilege1',
+  ].forEach(validValue => {
+    test(`returns true for '${validValue}'`, () => {
```
```suggestion
    test(`returns false for '${invalidValue}'`, () => {
```

legrego commented on x-pack/plugins/security/server/lib/authorization/privilege_serializer.test.ts:
```diff
@@ -36,6 +36,43 @@ describe(`#isSerializedSpaceBasePrivilege`, () => {
+    'bar',
+    'feature_foo',
+    'feature_foo.privilege1',
+  ].forEach(validValue => {
```
```suggestion
  ].forEach(invalidValue => {
```

legrego commented on x-pack/plugins/security/server/lib/authorization/privilege_serializer.test.ts:
```diff
@@ -36,6 +36,43 @@ describe(`#isSerializedSpaceBasePrivilege`, () => {
+    'feature_foo.privilege1',
+  ].forEach(validValue => {
+    test(`returns true for '${validValue}'`, () => {
+      expect(PrivilegeSerializer.isSerializedReservedPrivilege(validValue)).toBe(false);
```
```suggestion
      expect(PrivilegeSerializer.isSerializedReservedPrivilege(invalidValue)).toBe(false);
```

legrego commented on x-pack/plugins/security/server/lib/authorization/privilege_serializer.test.ts:
```diff
@@ -36,6 +36,43 @@ describe(`#isSerializedSpaceBasePrivilege`, () => {
+  });
+
+  ['all', 'read', 'space_all', 'space_read', 'reserved_foo', 'reserved_bar'].forEach(validValue => {
+    test(`returns true for '${validValue}'`, () => {
```
```suggestion
    test(`returns false for '${invalidValue}'`, () => {
```

legrego commented on x-pack/plugins/security/server/lib/authorization/privilege_serializer.test.ts:
```diff
@@ -36,6 +36,43 @@ describe(`#isSerializedSpaceBasePrivilege`, () => {
+    });
+  });
+
+  ['all', 'read', 'space_all', 'space_read', 'reserved_foo', 'reserved_bar'].forEach(validValue => {
```
```suggestion
  ['all', 'read', 'space_all', 'space_read', 'reserved_foo', 'reserved_bar'].forEach(invalidValue => {
```

legrego commented on x-pack/plugins/security/server/lib/authorization/privilege_serializer.test.ts:
```diff
@@ -36,6 +36,43 @@ describe(`#isSerializedSpaceBasePrivilege`, () => {
+
+  ['all', 'read', 'space_all', 'space_read', 'reserved_foo', 'reserved_bar'].forEach(validValue => {
+    test(`returns true for '${validValue}'`, () => {
+      expect(PrivilegeSerializer.isSerializedFeaturePrivilege(validValue)).toBe(false);
```
```suggestion
      expect(PrivilegeSerializer.isSerializedFeaturePrivilege(invalidValue)).toBe(false);
```

legrego commented on x-pack/plugins/security/server/routes/api/public/roles/get.js:
```diff
@@ -15,6 +15,34 @@ export function initGetRolesApi(server, callWithRequest, routePreCheckLicenseFn,
 
+    // if any application entry contains an empty resource, we throw an error
+    if (roleKibanaApplications.some(entry => entry.resources.length === 0)) {
+      throw new Error(`ES returned an application entry without resources, can't process this`);
```
Is there a reason we can't return `{ success: false }` for this case?

kobelb replied:
This really shouldn't ever happen, if it does Elasticsearch's validation rules have changed drastically. I chose to throw an Error explicitly here because the contract with ES has changed if this does occur, so it really is an unanticipated situation. The other transform errors are used when a user reaches around our APIs and writes directly to ES, so that's why these two things are being handled differently.

This error used to be being thrown below where we'd try to parse a feature incorrectly and the serializer would throw the error for us, but with the changes that I've made to support the reserved privileges, it was no longer hitting this flow so we're throwing this error explicitly.

legrego replied:
I that makes sense, thanks for explaining!

legrego commented on x-pack/test/functional/apps/apm/feature_controls/apm_spaces.ts:
```diff
@@ -0,0 +1,93 @@
+          basePath: '/s/custom_space',
+        });
+
+        await testSubjects.existOrFail('apm-main-container', 10000);
```
stray timeout
```suggestion
        await testSubjects.existOrFail('apm-main-container');
```

legrego committed:
Apply suggestions from code review

Co-Authored-By: kobelb <brandon.kobel@gmail.com>

kobelb committed:
Removing errant timeout

kobelb committed:
No more hard coded esFrom source

kobelb committed:
More nits

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/985/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/996/)

kobelb committed:
Adding back esFrom source

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1033/)

legrego reviewed (approved):
LGTM - I know we're waiting on a couple ES changes before this can merge, but I'll approve now to unblock you. Thanks for the edits, this looks great!

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

kobelb committed:
APM no longer uses reserved privileges, reserved privileges are
pluggable

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5681/)

kobelb committed:
Fixing typescript errors

kobelb committed:
Fixing ui capability test themselves

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5834/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5846/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5918/)

kobelb commented:
One of the previous deficiencies of this PR was that we weren't displaying the reserved privileges appropriately in the UI, but this has since been addressed. 

# Reserved Roles

When spaces is enabled, and the user views any of the reserved roles with a reserved privilege, a screen similar to the following will be displayed:

<img width="1254" alt="Screen Shot 2019-03-19 at 2 34 07 PM" src="https://user-images.githubusercontent.com/627123/54643468-36e20080-4a54-11e9-883a-9bf7bd6e618f.png">

And when spaces isn't enabled, the following will be displayed:

<img width="1255" alt="Screen Shot 2019-03-19 at 2 24 57 PM" src="https://user-images.githubusercontent.com/627123/54643424-187c0500-4a54-11e9-8c97-c6cab9b2e671.png">

# Custom Roles

We currently only plan to need these "reserved privileges" for Monitoring and Machine Learning's reserved roles, and I'm not aware of any situations where we recommend or require users to create their own custom roles for granting access to machine learning or monitoring (outside of Cloud which doesn't use any of the reserved roles). Nevertheless, the Role Management APIs have been written in such a way that it's possible to create a role with a reserved privilege if this ends up to not be the case. In these situations, I made changes so that we'll still display the reserved privileges on the forms, but not allow them to be edited.

When spaces is enabled, and they're editing a role with reserved privileges, they'll see a UI similar to the following:

![spaces-enabled-reserved-privileges](https://user-images.githubusercontent.com/627123/54643779-f767e400-4a54-11e9-820d-bf534d2a4316.gif)


And when spaces isn't enabled, the following will be displayed:

![spaces-disabled-reserved-privileges](https://user-images.githubusercontent.com/627123/54643981-7230ff00-4a55-11e9-9a74-cf1c69abb315.gif)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6078/)

kobelb committed:
Displaying reserved privileges for the space aware and simple forms

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6085/)

kobelb committed:
Removing ability to PUT roles with _reserved privileges.
Removing ability to GET roles that have entries with both reserved and
feature/base privileges.

kobelb commented:
One of the previous deficiencies of this PR was that we weren't displaying the reserved privileges appropriately in the UI, but this has since been addressed. 

# Reserved Roles

When spaces is enabled, and the user views any of the reserved roles with a reserved privilege, a screen similar to the following will be displayed:

<img width="1254" alt="Screen Shot 2019-03-19 at 2 34 07 PM" src="https://user-images.githubusercontent.com/627123/54643468-36e20080-4a54-11e9-883a-9bf7bd6e618f.png">

And when spaces isn't enabled, the following will be displayed:

<img width="1255" alt="Screen Shot 2019-03-19 at 2 24 57 PM" src="https://user-images.githubusercontent.com/627123/54643424-187c0500-4a54-11e9-8c97-c6cab9b2e671.png">

# Custom Roles

We currently only need these "reserved privileges" for Monitoring and Machine Learning's reserved roles. As such, the role management API has been changed to not allow roles to be created or updated which assign these reserved privileges. Additionally, if we come across an existing Role that has reserved privileges in addition to another privilege type, we won't transform this role.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6091/)

kobelb committed:
Updating jest snapshots

kobelb committed:
Changing the interface for a feature to register a reserved privilege to
include a description as well

kobelb committed:
Displaying features with reserved privileges in the feature table

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6500/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

kobelb committed:
Adjusting the reserved role privileges unit tests

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6524/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6541/)

chrisronline reviewed (approved):
This LGTM from a stack monitoring perspective. I tested spaces with and without monitoring enabled and the functionality is as expected!

spalger commented on x-pack/test/functional/apps/machine_learning/feature_controls/ml_security.ts:
```diff
@@ -0,0 +1,115 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import expect from 'expect.js';
```
Need to merge master and rewrite this import before merging: https://github.com/elastic/kibana/pull/33761

kobelb replied:
This PR is to merge it into the `granular-app-privileges` branch, we're gonna have to update a ton of these when doing so...

alvarezmelissa87 commented:
Tested this locally and everything is working as expected - LGTM 👍 😄

alvarezmelissa87 reviewed (approved):
Left a couple of comments but overall LGTM

peteharverson reviewed (approved):
ML functionality LGTM

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

kobelb committed:
Changing usages of expect.js to @kbn/expect

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7385/)

kobelb commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7426/)

legrego commented on x-pack/plugins/security/common/model/kibana_privileges/feature_privileges.ts:
```diff
@@ -19,7 +19,12 @@ export class KibanaFeaturePrivileges {
   public getPrivileges(featureId: string): string[] {
-    return Object.keys(this.featurePrivilegesMap[featureId]);
+    const featurePrivileges = this.featurePrivilegesMap[featureId];
+    if (featurePrivileges == null) {
```
thoughts on either `if (!featurePrivileges)` or `if (featurePrivileges === null)`? Here and elsewhere in this PR, I noticed a lot of loose equality checks (`==` as opposed to `===`). I know we've tried to limit using those in the past - is that still something we want to do?

kobelb replied:
In this situation, the `RawKibanaFeaturePrivileges` interface is the following:

```
interface RawKibanaFeaturePrivileges {
  [featureId: string]: {
    [privilegeId: string]: string[];
  };
}
```

so the two following are valid
```
const withUndefined: RawKibanaFeaturePrivileges = {
  something: undefined
};

const withNull: RawKibanaFeaturePrivileges = {
  something: null
};
```

I was using `== null` here to effectively check for either `null` or `undefined`, and if that's the situation return an empty array. IIRC, both ML and Monitoring are features and there's a code-path which tries to get the privileges for these features, but there are none, since they're using the special "reserved bag".

The other places which use `== null` should be in a similar situation where technically based on the types we could have either `null` or `undefined`. Doing the whole `if (!featurePrivileges)` felt wrong since it relies on the whole truthy/falsy values of JS.

legrego commented on x-pack/plugins/security/server/routes/api/public/roles/get.js:
```diff
@@ -15,13 +15,51 @@ export function initGetRolesApi(server, callWithRequest, routePreCheckLicenseFn,
     const roleKibanaApplications = roleApplications
```
We already discussed this, but we'll need changes here to support the `kibana-*` wildcard entry on the reserved role definitions.

kobelb commented on x-pack/plugins/security/public/lib/kibana_privilege_calculator/kibana_allowed_privileges_calculator.ts:
```diff
@@ -93,6 +93,10 @@ export class KibanaAllowedPrivilegesCalculator {
   ): { privileges: string[]; canUnassign: boolean } {
     const effectiveFeaturePrivilegeExplanation = effectivePrivileges.feature[featureId];
+    if (effectiveFeaturePrivilegeExplanation == null) {
+      throw new Error('To calculate allowed feature privileges, we need the effective privileges');
```
This code-path shouldn't be hit because the allowed privileges require the effective privileges to calculate, but the explicit Error should help us if this is no longer the case.

kobelb committed:
Changing the CalculatedPrivilege's _reserved property to reserved

kobelb commented on x-pack/plugins/security/public/lib/kibana_privilege_calculator/kibana_privilege_calculator_types.ts:
```diff
@@ -32,8 +32,9 @@ export interface PrivilegeExplanation {
-    [featureId: string]: PrivilegeExplanation;
+    [featureId: string]: PrivilegeExplanation | undefined;
   };
+  _reserved: undefined | string[];
```
I'm not opposed to it, I went back and forth on which one to use for the various usages. I will make it so.

kobelb commented on x-pack/plugins/xpack_main/server/lib/feature_registry/feature_registry.ts:
```diff
@@ -75,30 +98,15 @@ const schema = Joi.object({
+  }).required(),
   privilegesTooltip: Joi.string(),
+  reserved: Joi.object({
+    privilege: privilegeSchema.required(),
```
> We should augment the validateFeature function below to ensure that reserved privileges obey the same rules that normal privileges have to follow.
Good catch!
> Is there a valid scenario where a feature registers both reserved and "regular" privileges? What will happen in that case?
I don't think anything bad will happen if a feature does this in the future as I don't believe we need a feature to fit cleanly into either category. The only situation I can think of where we might use this is for transitioning an application like monitoring or machine learning to "Kibana Privileges". It didn't seem harmful to leave the flexibility there, but if you see a benefit to us locking this down, we can definitely do so.

legrego replied:
That makes sense. I was initially concerned that we could get [snagged by this check](https://github.com/elastic/kibana/pull/30525/files#diff-56fda9b9e6a451fa4247b3eb6c821522R46), but it's not possible to grant both reserved and non-reserved privileges to a role at this point, so we should be fine.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7722/)

kobelb committed:
Allowing reserved privileges to be assigned at kibana-*

kobelb committed:
Updating forgotten snapshot

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7745/)

kobelb commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7750/)

kobelb committed:
Validating reserved privileges

legrego reviewed (approved):
LGTM on green!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7861/)

kobelb commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7897/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8328/)

kobelb committed:
Updating imports

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8346/)

kobelb commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8365/)

kobelb committed:
Merge remote-tracking branch 'upstream/granular-app-privileges' into fc/reserved-role-apps-3

kobelb committed:
Removing --esFrom flag, we don't need it anymore

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8868/)

kobelb committed:
Switching from tslint's ignore to eslint's ignore

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8888/)
